subsys/bluetooth/host/adv.c

```diff
@@ -27,6 +27,107 @@ enum adv_name_type {
 	ADV_NAME_TYPE_SD,
 };
 
+struct bt_ad {
+	const struct bt_data *data;
+	size_t len;
+};
+
+struct ad_stream {
+	/* ad is a two dimensional array of struct bt_data elements. */
+	const struct bt_ad *ad;
+	size_t ad_len;
+
+	/* Indices into the two dimensional array of elements. */
+	size_t ad_index;
+	size_t data_index;
+
+	uint8_t current_data_offset;
+};
+
+void ad_stream_new(struct ad_stream *stream, const struct bt_ad *ad, size_t ad_len)
+{
+	memset(stream, 0, sizeof(*stream));
+	stream->ad = ad;
+	stream->ad_len = ad_len;
+}
+
+bool ad_stream_is_empty(const struct ad_stream *stream)
+{
+	return stream->ad_index == stream->ad_len;
+}
+
+/* Read data from the flattened AD stream. */
+uint8_t ad_stream_read(struct ad_stream *stream, uint8_t *buf, uint8_t *len)
+{
+	const struct bt_data *current_data = &stream->ad[stream->ad_index].data[stream->data_index];
+
+	const bool is_first_data = (stream->ad_index == 0) && (stream->data_index == 0);
+	const bool is_last_data = (stream->ad_index == stream->ad_len - 1) &&
+				  (stream->data_index == stream->ad[stream->ad_index].len - 1);
+
+	if (stream->current_data_offset == 0) {
+		/* Start reading out a new bt_data structure */
+
+		size_t max_data_to_be_read = current_data->data_len + 2;
+
+		buf[0] = current_data->data_len + 1;
+		buf[1] = current_data->type;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[2], current_data->data, max_data_to_be_read - 2);
+		stream->current_data_offset = max_data_to_be_read - 2;
+
+	} else {
+
+		/* Read the remaining data */
+		ssize_t max_data_to_be_read = current_data->data_len - stream->current_data_offset;
+
+		if (max_data_to_be_read <= *len) {
+			/* The entire bt_data structure fits within the buffer */
+			*len = max_data_to_be_read;
+		} else {
+			/* We can only fit parts of the data structure, the length is not updated */
+		}
+
+		(void)memcpy(&buf[0], current_data->data, max_data_to_be_read);
+		stream->current_data_offset += max_data_to_be_read;
+	}
+
+	uint8_t operation;
+
+	if (is_first_data && is_last_data &&
```

Comment:
I suppose that if `is_first_data && is_last_data` then it means that all data fit, right? 

`is_first_data && is_last_data` does read a bit weird, so we could consider storing that in a `const` variable as well (e.g. `is_complete_data`).

Reply:
We can't know if the data is complete before we have appended the data to the buffer. The data may be both first and last, but it is fragmented into multiple parts. Therefore we can't make it const before this evaluation. 

I can add a variable `is_complete_data` right before this if statement, but I'm not sure it will make the code more readable.

Reply:
Ignore this comment; it was based on an incorrect understanding of what `data` meant here.